examples/demo-app/src/cloud-providers/aws/aws-provider.js

```diff
@@ -0,0 +1,347 @@
+    const mapFileList = fileList.filter(file => file.key.endsWith('map.json'));
+
+    const updatedFileList = mapFileList.map(file => {
+      const title = file.key.split('.')[0];
+      const thumbnailKey = `${title}.thumbnail.png`;
+      const metaKey = `${title}.meta.json`;
+
+      const fileData = {
+        id: file.key,
+        title,
+        description: '',
+        privateMap: level === 'private',
+        thumbnail: '',
+        lastModification: new Date(Date.parse(file.lastModified)),
+        loadParams: {
+          identityId: '',
+          mapId: file.key,
+          level
+        }
+      };
+
+      const thumbnailUrl = fileList.some(f => f.key === thumbnailKey)
+        ? this._getFile(thumbnailKey, 'image', {level, download: false})
+        : null;
+      const description = fileList.some(f => f.key === metaKey)
+        ? this._getFile(metaKey, 'meta data', {level, download: true})
+        : 'No description meta data file';
+
+      return Promise.all([thumbnailUrl, description])
+        .then(respList => {
+          fileData.thumbnail = respList[0];
+          fileData.description = respList[1];
+        })
+        .then(resp => fileData)
+        .catch(e => this._handleError(e));
+    });
+    return Promise.all(updatedFileList);
+  }
+
+  /**
+   *
+   * @returns {MapResponse}
+   */
+  async downloadMap(loadParams) {
+    const {level, mapId, identityId} = loadParams;
+    return Storage.get(mapId, {level, ...(level === 'private' ? {} : {identityId})})
+      .then(fetch)
+      .then(response => response.json())
+      .then(mapData => {
+        if (this._loadParam !== loadParams) {
+          this._loadParam = loadParams;
+        }
+        return {
+          map: mapData,
+          format: 'keplergl'
+        };
+      })
+      .catch(e => AwsProvider._handleError(`Map downloading failed`, e));
+  }
+
+  /**
+   * Save if isPublic false or
+   * ShareUrl if isPublic true
+   * @returns {Promise<{level, mapId, identityId} || {shareUrl}>}
+   * You can share url with saved map through public map link (as defined)
+   * or through loadParams used in downloadMap (set SHARING_WITH_MAP_URL to false to use)
+   * in second case, the user has to be logged in to open the map
+   */
+  async uploadMap({mapData, options = {}}) {
+    const {isPublic} = options;
+    const {map, thumbnail} = mapData;
+    const {title, description} = map && map.info;
+    const name = title;
+    // Since we share through a map link, this could be private as well
+    const level = isPublic ? 'protected' : 'private';
+    const saveThumbnail = this._saveFile(name, 'thumbnail.png', thumbnail, level);
+    const saveMeta = this._saveFile(name, 'meta.json', {description}, level);
+    const saveMap = this._saveFile(name, 'map.json', map, level);
+
+    return Promise.all([saveThumbnail, saveMeta, saveMap])
+      .then(savedList => {
+        const key = savedList && savedList[2] && savedList[2].key;
```

Comment:
@rollacaster 
Maybe here it would also be better to do it analog as above? 

    return Promise.all([saveThumbnail, saveMeta, saveMap])
      .then(([thumbnailSaved, metaSaved, mapSaved]) => {
        const key = mapSaved && mapSaved.key;
...

Reply:
yes 👍